src/c_string.rs

```diff
@@ -0,0 +1,469 @@
+    /// ```
+    pub unsafe fn from_bytes_with_nul_unchecked(bytes: &[u8]) -> Result<Self, CapacityError> {
+        let mut vec = Vec::new();
+
+        vec.extend_from_slice(bytes)?;
+
+        Ok(Self { vec })
+    }
+
+    /// Instantiate a [`CString`] copying from the giving bytes, assuming it is
+    /// null-terminated (ends with a single zero byte).
+    ///
+    /// Fails if the given byte slice has any interior null byte, if the slice does not
+    /// end in a zero byte or if the byte slice can't fit in `CAP`.
+    pub fn from_bytes_with_nul(bytes: &[u8]) -> Result<Self, CapacityError> {
+        let mut me = Self::new();
+
+        me.push_bytes(bytes)?;
+
+        Ok(me)
+    }
+
+    /// Builds a [`CString`] copying from a raw C string pointer.
+    ///
+    /// # Safety
+    ///
+    /// * The memory pointed to by `ptr` must contain a valid null terminator at the
+    ///   end of the string.
+    ///
+    /// * `ptr` must be valid for reads of bytes up to and including the null terminator.
+    ///   This means in particular:
+    ///
+    ///     * The entire memory range of this `CStr` must be contained within a single allocated object!
+    ///     * `ptr` must be non-null even for a zero-length cstr.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    /// use std::ffi::{c_char, CStr};
+    ///
+    /// const HELLO_PTR: *const c_char = {
+    ///     const BYTES: &[u8] = b"Hello, world!\0";
+    ///     BYTES.as_ptr().cast()
+    /// };
+    ///
+    /// let copied = unsafe { CString::<14>::from_ptr(HELLO_PTR) }.unwrap();
+    ///
+    /// assert_eq!(copied.to_str(), Ok("Hello, world!"));
+    /// ```
+    pub unsafe fn from_ptr<'a>(ptr: *const c_char) -> Result<Self, CapacityError> {
+        let cstr = CStr::from_ptr(ptr).to_bytes_with_nul();
+
+        Self::from_bytes_with_nul(cstr)
+    }
+
+    /// Converts the [`CString`] to a [`CStr`] slice. This always succeeds and is zero cost.
+    pub fn as_c_str(&self) -> &CStr {
+        debug_assert!(CStr::from_bytes_with_nul(&self.vec).is_ok());
+
+        unsafe { CStr::from_bytes_with_nul_unchecked(&self.vec) }
+    }
+
+    /// How many bytes were inserted to this [`CString`] so far, considering its
+    /// null terminator.
+    ///
+    /// Must always be bigger than zero, since even an empty [`CString`]
+    /// ends in a zero byte.
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    /// use std::ffi::{c_char, CStr};
+    ///
+    /// // Length is one (null terminator only), capacity is 10
+    /// let mut cstr = CString::<11>::new();
+    ///
+    /// // Insert 5 bytes to it
+    /// cstr.push_bytes(b"/etc/").unwrap();
+    /// // Length is 6 (5 bytes inserted plus null terminator)
+    /// assert_eq!(cstr.len(), 6);
+    /// cstr.push_bytes(b"dconf").unwrap();
+    ///
+    /// assert_eq!(cstr.to_str(), Ok("/etc/dconf"));
+    /// ```
+    pub fn len(&self) -> usize {
```

Comment:
Not sure about this as `std::ffi::CString` didn't have this. To me, it looks like they didn't add it intentionally in order to avoid confusion around the null terminator...

Reply:
I removed it for now. In the future, if the Rust's `std library team adds it to `CStr`, it will be available via `Deref`.

Reply:
A more general question is if we need these utility functions implemented for `heapless::CString` which should technically be implemented for the upstream `core::ffi::CStr`.

Reply:
Makes sense to drop `len`, since `CStr::count_bytes` will be available from Deref